src/bit-systems/audio-emitter-system.ts

```diff
@@ -0,0 +1,113 @@
+}
+
+export function cleanupAudio(audio: AudioObject3D) {
+  const eid = audio.eid!;
+  audio.disconnect();
+  const audioSystem = APP.scene?.systems["hubs-systems"].audioSystem;
+  APP.audios.delete(eid);
+  APP.supplementaryAttenuation.delete(eid);
+  APP.audioOverrides.delete(eid);
+  audioSystem.removeAudio({ node: audio });
+}
+
+function swapAudioType<T extends AudioObject3D>(
+  world: HubsWorld,
+  audioSystem: AudioSystem,
+  eid: number,
+  NewType: AudioConstructor<T>
+) {
+  const audio = world.eid2obj.get(eid)! as AudioObject3D;
+  audio.disconnect();
+  APP.gainMultipliers.delete(eid);
+  APP.sourceType.set(eid, SourceType.MEDIA_VIDEO);
+  APP.supplementaryAttenuation.delete(eid);
+  APP.audioOverrides.delete(eid);
+  APP.audios.delete(eid);
+  audioSystem.removeAudio({ node: audio });
+
+  const newAudio = new NewType(APP.audioListener);
+  newAudio.setNodeSource(audio.source!);
+  audioSystem.addAudio({ sourceType: SourceType.MEDIA_VIDEO, node: newAudio });
+  APP.audios.set(eid, newAudio);
+
+  audio.parent!.add(newAudio);
+  audio.removeFromParent();
+
+  swapObject3DComponent(world, eid, newAudio);
+}
+
+export function makeAudioSourceEntity(world: HubsWorld, video: HTMLVideoElement, audioSystem: AudioSystem) {
+  const eid = addEntity(world);
+  APP.sourceType.set(eid, SourceType.MEDIA_VIDEO);
+  if (video.paused) {
+    AudioEmitter.flags[eid] |= EMITTER_FLAGS.PAUSED;
+  } else {
+    AudioEmitter.flags[eid] &= ~EMITTER_FLAGS.PAUSED;
+  }
+
+  let audio;
+  const { audioType } = getCurrentAudioSettings(eid);
+  const audioListener = APP.audioListener;
+  if (audioType === AudioType.PannerNode) {
+    audio = new PositionalAudio(audioListener);
+  } else {
+    audio = new StereoAudio(audioListener);
+  }
+  addComponent(world, AudioEmitter, eid);
+  addObject3DComponent(world, eid, audio);
+
+  audio.gain.gain.value = 0;
+  audioSystem.addAudio({ sourceType: SourceType.MEDIA_VIDEO, node: audio });
+
+  const audioSrcEl = video;
+  audio.setMediaElementSource(audioSrcEl);
+
+  APP.audios.set(eid, audio);
+  updateAudioSettings(eid, audio);
+
+  // Original audio source volume can now be restored as audio systems will take over
+  audioSrcEl.volume = 1;
+  return eid;
+}
+
+const staleAudioEmittersQuery = defineQuery([AudioEmitter, AudioSettingsChanged]);
+export function audioEmitterSystem(world: HubsWorld, audioSystem: AudioSystem) {
+  staleAudioEmittersQuery(world).forEach(function (eid) {
+    const audio = world.eid2obj.get(eid)! as PositionalAudio | StereoAudio;
+    const settings = getCurrentAudioSettings(eid);
+    const isPannerNode = isPositionalAudio(audio);
+
+    // TODO this needs more testing
+    if (!isPannerNode && settings.audioType === AudioType.PannerNode) {
+      swapAudioType(world, audioSystem, eid, PositionalAudio);
+    } else if (isPannerNode && settings.audioType === AudioType.Stereo) {
+      swapAudioType(world, audioSystem, eid, StereoAudio);
+    }
+
+    applySettings(audio, settings);
```

Comment:
If the settings can change during `swapAudioType`, then we should `getCurrentAudioSettings` after we (maybe) swap.

Reply:
Not sure... I'd rather leave this as is and keep an eye on it rather than doing defensive coding.